backend/src/org/akvo/dash/transformation.clj

```diff
@@ -1,225 +1,157 @@
+                    (on-error-set onError)))
+      (throw-invalid-op op-spec)))
+
+(defmulti valid-op?
+  (fn [op-spec]
+    (keyword (op-spec "op"))))
+
+(defmethod valid-op? :default
+  [op-spec]
+  (throw-invalid-op op-spec))
+
+(defmethod valid-op? :core/change-datatype
+  [{:strs [op args] :as op-spec}]
+  (or (boolean (and (required-keys op-spec)
+                 (type-set (args "newType"))
+                 (not-empty (args "defaultValue"))
+                 (if (= "date" (args "newType"))
+                   (not-empty (args "parseFormat"))
+                   true)))
+      (throw-invalid-op op-spec)))
+
+(defmethod valid-op? :core/change-column-title
+  [{:strs [args] :as op-spec}]
+  (or (boolean (and (required-keys op-spec)
+                    (not-empty (args "columnTitle"))))
+      (throw-invalid-op op-spec)))
+
+(defmethod valid-op? :core/sort-column
+  [{:strs [args] :as op-spec}]
+  (or (boolean (and (required-keys op-spec)
+                    (sort-direction-set (args "sortDirection"))))
+      (throw-invalid-op op-spec)))
+
+(defmethod valid-op? :core/remove-sort
+  [op-spec]
+  (or (required-keys op-spec)
+      (throw-invalid-op op-spec)))
+
+(defmethod valid-op? :core/filter
+  [op-spec]
+  true)
+
+(defmethod valid-op? :core/to-titlecase
+  [op-spec]
+  (required-keys op-spec))
+
+(defmethod valid-op? :core/to-lowercase
+  [op-spec]
+  (or (required-keys op-spec)
+      (throw-invalid-op op-spec)))
+
+(defmethod valid-op? :core/to-uppercase
+  [op-spec]
+  (or (required-keys op-spec)
+      (throw-invalid-op op-spec)))
+
+(defmethod valid-op? :core/trim
+  [op-spec]
+  (or (required-keys op-spec)
+      (throw-invalid-op op-spec)))
+
+(defmethod valid-op? :core/trim-doublespace
+  [op-spec]
+  (or (required-keys op-spec)
+      (throw-invalid-op op-spec)))
+
+(defn valid?
+  [transformations]
+  (try
+    {:valid? (every? valid-op? transformations)}
+    (catch Exception e
+      {:valid? false
+       :message (.getMessage e)})))
+
+(defn execute
+  [tenant-conn job-id dataset-id transformations]
+  (let [dv (dataset-version-by-id tenant-conn {:id dataset-id})
+        columns (vec (:columns dv))
+        source-table (:imported-table-name dv)
+        imported-table (gen-table-name "imported")
+        table-name (gen-table-name "ds")
+        f (fn [log op-spec]
+            (let [step (engine/apply-operation tenant-conn
+                                               table-name
+                                               columns
+                                               op-spec)]
+              (if (:success? step)
+                (conj log step)
+                (throw (Exception. (str "Error applying operation: " op-spec))))))]
+    (try
+      (copy-table tenant-conn {:source-table source-table
+                               :dest-table imported-table})
+      (copy-table tenant-conn {:source-table imported-table
+                               :dest-table table-name})
```

Comment:
We have a unique constraint in the `dataset_version` table that doesn't allow us to have the same `dataset_version` with the same `imported_table_name` (a waste of space!). We can remove this double cloning and more than one `dataset_version` can depend on the same `imported_table_name`